sklearn/cluster/optics_.py

```diff
@@ -0,0 +1,769 @@
+            self._processed[point] = True
+
+    def _set_reach_dist(self, point_index, X):
+        P = np.array(X[point_index]).reshape(1, -1)
+        indices = self.radius_neighbors(P, radius=self.max_bound,
+                                        return_distance=False)[0]
+
+        # Checks to see if there more than one member in the neighborhood
+        if iterable(indices):
+            # Masking processed values; n_pr is 'not processed'
+            n_pr = np.compress((np.take(self._processed,
+                                        indices, axis=0) < 1).ravel(),
+                               indices, axis=0)
+            # n_pr = indices[(self._processed[indices] < 1).ravel()]
+            # Keep n_jobs = 1 in the following lines...please
+            if len(n_pr) > 0:
+                dists = pairwise_distances(P, np.take(X, n_pr, axis=0),
+                                           self.metric, n_jobs=1).ravel()
+
+                rdists = np.maximum(dists, self.core_dists_[point_index])
+                new_reach = np.minimum(np.take(self.reachability_,
+                                               n_pr, axis=0), rdists)
+                self.reachability_[n_pr] = new_reach
+
+            # Checks to see if everything is already processed;
+            # if so, return control to main loop
+            if n_pr.size > 0:
+                # Define return order based on reachability distance
+                return(n_pr[quick_scan(np.take(self.reachability_,
+                                               n_pr, axis=0), dists)])
+            else:
+                return point_index
+
+    def extract_dbscan(self, eps):
+        """Performs DBSCAN extraction for an arbitrary epsilon.
+
+        Parameters
+        ----------
+        eps: float or int, required
+            DBSCAN `eps` parameter. Must be set to < `max_bound`. Equivalence
+            with DBSCAN algorithm is achieved if `eps` is < (`max_bound` * 5)
+
+        Returns
+        -------
+        core_sample_indices_: array, shape (n_core_samples,)
+            The indices of the core samples.
+
+        labels_ : array, shape (n_samples,)
+            The estimated labels.
+        """
+        check_is_fitted(self, 'reachability_')
+
+        if eps > self.max_bound:
+            raise ValueError('Specify an epsilon smaller than %s. Got %s.'
+                             % (self.max_bound, eps))
+
+        if eps * 5.0 > (self.max_bound * 1.05):
+            warnings.warn(
+                "Warning, max_bound (%s) is close to eps (%s): "
+                "Output may be unstable." % (self.max_bound, eps),
+                RuntimeWarning, stacklevel=2)
+        # Stability warning is documented in extract_dbscan method...
+
+        return extract_dbscan(self.ordering_, self.core_dists_,
+                              self.reachability_, eps)
+
+
+def extract_dbscan(ordering, core_dists, reachability, eps):
```

Comment:
this public function is never tested alone nor is it documented. Please make it private. extract_dbscan -> _extract_dbscan

Reply:
ok, makes sense-- fixed